### PR TITLE
[FIX] Combine positive and negative cluster labels for two sided permutation tests 

### PR DESCRIPTION
**Pull request by jrdalenberg**
This addresses the bug report in #3435.

### CODE REVIEW

File: nilearn/mass_univariate/permuted_least_squares.py [hunk 1]
```diff
@@ -903,15 +903,13 @@ def permuted_ols(
...
             if two_sided_test:
-                # Label positive and negative clusters separately
-                n_positive_clusters = np.max(labeled_arr3d)
+                # Add negative cluster labels
                 temp_labeled_arr3d, _ = label(
                     scores_original_data_3d < -threshold_t,
                     bin_struct,
                 )
-                temp_labeled_arr3d[
-                    temp_labeled_arr3d > threshold_t
```
Comment (tsalo) [inline nilearn/mass_univariate/permuted_least_squares.py:913]: Would the bug be fixed by changing `threshold_t` here to `0`?

**Review by bthirion — COMMENTED** [review]
Thx ! Could you add a test that fails with previous version and passes with the fix ?

File: nilearn/mass_univariate/permuted_least_squares.py (hunk 1 above)
Comment (jrdalenberg) [reply nilearn/mass_univariate/permuted_least_squares.py:913]: Yes, you are right. I did not fully understand the logic at first and reversed the order it seems: 

`temp_labeled_arr3d > 0` results in `np.concatenate([positive cluster labels], [negative cluster labels])`

And my change results in `np.concatenate([negative clusters labels], [positive cluster labels])`

Which do you prefer?

**Comment by jrdalenberg** [comment]
> Thx ! Could you add a test that fails with previous version and passes with the fix ?

Sure! But I first need to figure out how to build tests 😁.

If you want to have a look already:

```python

from nilearn.datasets import fetch_localizer_contrasts
from nilearn.glm.second_level import non_parametric_inference
from nilearn.glm.second_level import SecondLevelModel
import pandas as pd


n_subjects = 16
data = fetch_localizer_contrasts(
    ['left vs right button press'],
    n_subjects,
    get_tmaps=True,
    legacy_format=False,
)

second_level_input = data['cmaps']
design_matrix = pd.DataFrame(
    [1] * len(second_level_input),
    columns=['intercept'],
)

second_level_model = SecondLevelModel(smoothing_fwhm=8.0)
second_level_model = second_level_model.fit(
    second_level_input,
    design_matrix=design_matrix,
)

out_dict = non_parametric_inference(
    second_level_input,
    design_matrix=design_matrix,
    model_intercept=True,
    n_perm=500,  # 500 for the sake of time. Ideally, this should be 10,000.
    two_sided_test=True,
    smoothing_fwhm=8.0,
    n_jobs=8,
    threshold=0.001,
)
```

This fails with the old code. 

(I borrowed the example from:
https://nilearn.github.io/dev/auto_examples/05_glm_second_level/plot_second_level_one_sample_test.html)

**Comment by ymzayek** [comment]
@jrdalenberg for some guidance you take a look at already written tests. Perhaps you can just add to one of them or copy one but with specific parameters to reproduce your failure.

For example see: https://github.com/nilearn/nilearn/blob/main/nilearn/mass_univariate/tests/test_permuted_least_squares.py#L408-L470
(Or if it needs to be tested on a higher level: https://github.com/nilearn/nilearn/blob/main/nilearn/glm/tests/test_second_level.py)

For tests we use fake generated data. There are examples in those test files you can copy. Let us know if you need any further help!

File: nilearn/mass_univariate/permuted_least_squares.py (hunk 1 above)
Comment (tsalo) [reply nilearn/mass_univariate/permuted_least_squares.py:913]: I think the diff would be cleaner if you just changed the threshold, but I don't feel too strongly about it. I don't know if changing the order of clusters in the labeled map (negative before positive, instead of the opposite) is something users will notice, so it's up to you.

**Comment by jrdalenberg** [comment]
I updated the `test_cluster_level_parameters_smoke` function. The test needed a sufficient amount of clusters for it to fail on `two_sided_test=True`.

**Comment by bthirion** [comment]
Can you check the Python 3.9 failure ? Seems related to the commit ...

**Comment by jrdalenberg** [comment]
> Can you check the Python 3.9 failure ? Seems related to the commit ...

The test crashed on:
https://github.com/nilearn/nilearn/blob/4d3208a13062a908450fc5dc71905ae7ce893a64/nilearn/mass_univariate/permuted_least_squares.py#L921

The log file does not show the entire `target_vars` variable:

```python
2022-12-02T15:24:45.4949550Z target_vars = array([[ 0.2658434 ,  0.2658434 ,  0.2658434 , ...,  0.2658434 ,
2022-12-02T15:24:45.4949780Z          0.2658434 ,  0.        ],
2022-12-02T15:24:45.4950130Z        [ 0.0520910...405, -8.        ],
2022-12-02T15:24:45.4950350Z        [ 0.88161576,  0.88161576,  0.88161576, ...,  0.88161576,
2022-12-02T15:24:45.4950650Z          0.88161576, -9.        ]])
```

So it is hard to replicate what it used as input. 

I tried to troubleshoot it locally on a Macbook 2017 running macOS Montery (12.6), but the test runs fine on python 3.9 and 3.10. 

Any help would be very welcome.

**Comment by ymzayek** [comment]
I'll have a look but I think this could be a random failure. @jrdalenberg see https://nilearn.github.io/stable/development.html#tests the section about seeding to see if you can improve the test.

**Comment by ymzayek** [comment]
@jrdalenberg thanks for making an update. I'll do a thorough review of the test as soon as I can. I just want to quickly point out that the failure in the Ubuntu-latest Python 3.10 pre-release environment is not related to this PR so you don't need to worry about it.

File: nilearn/mass_univariate/tests/test_permuted_least_squares.py [hunk 2]
```diff
@@ -595,19 +595,27 @@ def test_tfce_smoke(random_state=0):
 
 def test_cluster_level_parameters_smoke(random_state=0):
     """Test combinations of parameters related to cluster-level inference."""
+    import random
```
Comment (ymzayek) [inline nilearn/mass_univariate/tests/test_permuted_least_squares.py:598]:
```suggestion
```

File: nilearn/mass_univariate/tests/test_permuted_least_squares.py [hunk 3]
```diff
@@ -595,19 +595,27 @@ def test_tfce_smoke(random_state=0):
...
     import nibabel as nib
     from nilearn.maskers import NiftiMasker
 
+    random.seed(random_state)
+    np.random.seed(random_state)
```
Comment (ymzayek) [inline nilearn/mass_univariate/tests/test_permuted_least_squares.py:603]:
```suggestion
    rng = np.random.RandomState(0)
```

File: nilearn/mass_univariate/tests/test_permuted_least_squares.py [hunk 4]
```diff
@@ -595,19 +595,27 @@ def test_tfce_smoke(random_state=0):
...
+
     # create design
     target_var1 = np.arange(0, 10).reshape((-1, 1))  # positive effect
-    target_var = np.hstack((  # corresponds to 3 x 3 x 3 x 10 niimg
-        target_var1,  # voxel 1 has positive effect
-        - target_var1,  # voxel 2 has negative effect
-        np.random.random((10, 25)),  # 25 remaining voxels
+    voxel_vars = np.hstack((
```
Comment (ymzayek) [inline nilearn/mass_univariate/tests/test_permuted_least_squares.py:607]:
```suggestion
    voxel_vars = np.hstack((
```

File: nilearn/mass_univariate/permuted_least_squares.py [hunk 5]
```diff
@@ -903,15 +903,13 @@ def permuted_ols(
...
-                ] += n_positive_clusters
+                n_negative_clusters = np.max(temp_labeled_arr3d)
```
Comment (ymzayek) [inline nilearn/mass_univariate/permuted_least_squares.py:911]:
```suggestion
                n_negative_clusters = np.max(temp_labeled_arr3d)
```

**Review by ymzayek — COMMENTED** [review]
@jrdalenberg I made some suggestions below. Some are just PEP8 related fixes. For the test I think it's best to stick to one generator so I replaced Python's random and I use a numpy.random.RandomState() instance instead of the numpy.random.seed function

File: nilearn/mass_univariate/tests/test_permuted_least_squares.py (hunk 3 above)
Comment (jrdalenberg) [reply nilearn/mass_univariate/tests/test_permuted_least_squares.py:603]: Oh cool! I did not know about this and that the Python stdlib module “random” is aware of this.

Q: Shouldn't this state be inherited from the function input `random_state` instead of hard coding it to 0?

Comment (ymzayek) [reply nilearn/mass_univariate/tests/test_permuted_least_squares.py:603]: @jrdalenberg yes my mistake,  it should take `random_state` as an argument

**Review by ymzayek — COMMENTED** [review]
Thanks @jrdalenberg for the timely changes. Something a bit funny happened in my last review and not all my suggestions seem to have been recorded. Below are mainly PEP8 fixes. Otherwise LGTM! 
As a first-time contributor can you add your name to `doc/changes/names.rst` and then add a whatsnew entry to `doc/changes/latest.rst`?

**Comment by ymzayek** [comment]
@jrdalenberg can you rebase on main and resolve the conflict? Then it should be ready to merge. Thanks!

**Comment by jrdalenberg** [comment]
@ymzayek done :)

**Review by ymzayek — APPROVED** [review]
@jrdalenberg thanks! I think this one can be merged. The test failures are not related to this PR. They are the same as reported in #3441

**Comment by ymzayek** [comment]
Merging

**Comment by bthirion** [comment]
Thx !
